runtime/src/crowdfund.rs

```diff
@@ -0,0 +1,539 @@
+				ExistenceRequirement::AllowDeath
+			)?);
+
+			T::OrphanedFunds::on_unbalanced(T::Currency::withdraw(
+				&account,
+				fund.raised,
+				WithdrawReason::Transfer,
+				ExistenceRequirement::AllowDeath
+			)?);
+
+			let id = Self::id_from_index(index);
+			child::kill_storage(id.as_ref());
+			<Funds<T>>::remove(index);
+
+			Self::deposit_event(RawEvent::Dissolved(index));
+
+		}
+		
+		
+		/// Set the deploy data of the funded parachain if not already set. Once set, this cannot
+		/// be changed again.
+		///
+		/// - `origin` must be the fund owner.
+		/// - `index` is the fund index that `origin` owns and whose deploy data will be set.
+		/// - `code_hash` is the hash of the parachain's Wasm validation function.
+		/// - `initial_head_data` is the parachain's initial head data.
+		fn fix_deploy_data(
+			origin,
+			#[compact] index: FundIndex,
+			code_hash: T::Hash,
+			initial_head_data: Vec<u8>
+		) {
+			let who = ensure_signed(origin)?;
+
+			let mut fund = Self::funds(index).ok_or("invalid fund index")?;
+			ensure!(fund.owner == who, "origin must be fund owner");
+			ensure!(fund.deploy_data.is_none(), "deploy data already set");
+
+			fund.deploy_data = Some((code_hash, initial_head_data));
+
+			<Funds<T>>::insert(index, &fund);
+
+			Self::deposit_event(RawEvent::DeployDataFixed(index));
+
+		}
+		
+		/// Complete onboarding process for a winning parachain fund. This can be called once by
+		/// any origin once a fund wins a slot and the fund has set its deploy data (using
+		/// `fix_deploy_data`).
+		///
+		/// - `index` is the fund index that `origin` owns and whose deploy data will be set.
+		/// - `para_id` is the parachain index that this fund won.
+		fn onboard(
+			_origin,
+			#[compact] index: FundIndex,
+			#[compact] para_id: ParaIdOf<T>
+		) {
+			let mut fund = Self::funds(index).ok_or("invalid fund index")?;
+			let (code_hash, initial_head_data) = fund.clone().deploy_data.ok_or("deploy data not fixed")?;
+			ensure!(fund.parachain.is_none(), "fund already onboarded");
+			fund.parachain = Some(para_id);
+
+			let fund_origin = system::RawOrigin::Signed(Self::fund_account_id(index)).into();
+			<slots::Module<T>>::fix_deploy_data(fund_origin, index, para_id, code_hash, initial_head_data)?;
+
+			<Funds<T>>::insert(index, &fund);
+
+			Self::deposit_event(RawEvent::Onboarded(index, para_id));
+		}
+		
+		fn on_finalize(n: T::BlockNumber) {
+			if <slots::Module<T>>::is_ending(n).is_some() {
+				for (fund, index) in NewRaise::take().into_iter().filter_map(|i| Self::funds(i).map(|f| (f, i)))
+				{
+					if fund.last_contribution == Some(n) {
+						let bidder = slots::Bidder::New(slots::NewBidder {
+							who: Self::fund_account_id(index),
+							/// FundIndex and slots::SubId happen to be the same type (u32). If this
+							/// ever changes, then some sort of conversion will be needed here.
+							sub: 0,
+						});
+						// TODO for review: Can this fail? Anything we can do to handle possible errors?
```

Comment:
Some comments above fix other possible error conditions. One more place it can fail is:

```
			// This must overlap with all existing ranges that we're winning on or it's invalid.
			ensure!(current_winning.iter()
				.enumerate()
				.all(|(i, x)| x.as_ref().map_or(true, |(w, _)|
					w != &bidder || range.intersects(i.try_into()
						.expect("array has SLOT_RANGE_COUNT items; index never reaches that value; qed")
					)
				)),
				"bidder winning non-intersecting range"
			);
```

Should we test for this somewhere throughout this auction process? is it okay if this fails?

Reply:
it's not actually possible for it to fail in this way since each fund is a unique bidder. but yeah, it's not a big problem if it fails here in principle (except for the individuals involved, but they should have been more careful).